Refine json comment position for Utf8JsonWriter; Make WriteCommentValue() overloads have same behavior

lateapexearlyspeed opened the pull request:
Fix #52408 and #75075 
Will add related tests after deciding correct format, thanks !

krwq commented on src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Comment.cs:
```diff
@@ -207,12 +207,10 @@ private void WriteCommentIndented(ReadOnlySpan<byte> utf8Value)
-                {
-                    WriteNewLine(output);
-                }
+                WriteNewLine(output);
```
please add tests (even if throws exception), at minimum:
- testing behavior you added,
- comment in the beginning of file
- comment in the end of file
- comment in the beginning of the object
- comment in the middle of the object
- comment as last element of object
- same as above but array rather than object

lateapexearlyspeed replied:
Test added.

krwq reviewed (commented):
Please also edit PR description to describe breaking change (diff and short comment is enough). Assuming this gets merged we'll need issue created as per instruction above

lateapexearlyspeed commented:
Hi @krwq just confirm, I will change code to implement just [proposal description](https://github.com/dotnet/runtime/issues/52408#issue-877928403) and align one [overload behavior issue](https://github.com/dotnet/runtime/issues/75075#issue-1361510964), and keep all other existing behavior no change.

I have pushed one version of prod code, and will continue adding tests later. If possible, maybe you can review and point out issue in advance, thanks !

lateapexearlyspeed commented:
Hi @krwq Tests added, please review, thanks!

They used newtonsoft.json to do assertion as existing cases, but have to add some compensation newlines and whitespaces so that can verify new cases. Note the compensations are not entirely for new proposal, partly it is because we lost some test cases for existing indented behaviors of comment previously.

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:
```diff
@@ -4067,9 +4068,9 @@ public void WriteInvalidDepthPartial(bool formatted, bool skipValidation)
         [InlineData(false, true, ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")]
         [InlineData(false, false, ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")]
-        public void WriteComments(bool formatted, bool skipValidation, string comment)
+        public void WriteCommentsInArray(bool formatted, bool skipValidation, string comment)
```
Really hard to tell just by looking at these tests, what format do we end up keeping in #75075?

lateapexearlyspeed replied:
For #75075, because it was closed without decision about format, so I picked this format based on [comment](https://github.com/dotnet/runtime/issues/75075#issue-1361510964):
> I just guess current difference is not intentional. If we should make them same, which one should we pick ?
> 
> I updated code to make them same as this one, please decide if another is better.
> 
> ```
> {
>   "property": 
>   /*comment*/"value"
> ```

Which is existing format of string parameter overload of `WriteCommentValue()`, and because it is no matter which type of property value, so this format should be more align with #52408 

Note this PR is to fix both #75075 and #52408, I just updated PR description now.

eiriktsarpalis replied:
Hmm, on closer inspection both formats seem kind of strange to me, especially if the comment value contains line breaks. Could this possibly be
```json
{
    "property":
    /*comment*/
    "value"
}
```
But that's just me. Do we have any examples of other (non .NET) JSON formatters that handle comments?

@krwq @layomia opinions?

eiriktsarpalis replied:
I see that this differs depending on whether the property value is a string or object. What is the exact rule being used here?

lateapexearlyspeed replied:
> I see from the tests that this differs depending on whether the property value is a string or object. What is the exact rule being used here?

The differ you found is just because current PR is to fix #52408. Current used rule is also based on those proposals, which is:

- for object and array: insert newlines between comment and object (or array), which is exactly described by #52408
- for other token types (not only string): keep existing behaviors not break, because #52408 doesn't  proposal them.

#75075 is to align 2 WriteCommentValue overload's different behavior about whether add newline between property name and comment, or not add newline. It is not related to property value currently because overload's difference is not there.

Current PR code should have been implemented based on rule above.

krwq replied:
@eiriktsarpalis what you proposed makes sense to me but also:

```json
{
    "property": /*comment*/
    "value"
}
```

would look good to me, I'm not strongly opinionated here - it seems like a weird place to put comment to me anyway

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:
```diff
@@ -4067,9 +4068,9 @@ public void WriteInvalidDepthPartial(bool formatted, bool skipValidation)
+        public void WriteCommentsInArray(bool formatted, bool skipValidation, string comment)
         {
-            string expectedStr = GetCommentExpectedString(prettyPrint: formatted, comment);
+            string expectedStr = GetCommentInArrayExpectedString(prettyPrint: formatted, comment);
```
Per my comment above, would it be possible to include a couple of smoke tests that check indentation against a couple of string literals rather than constructing the expected string using test logic? Both might contain bugs or could regress.

krwq replied:
Agreed, I'm looking at the test and it's really hard to figure out at first glance what this test is actually testing. I'd prefer couple of hardcoded string cases as well

lateapexearlyspeed replied:
Even I agree :) Changed to use string literal now.
Just note in advance when you view inside string literal result: 

> /*Comment in the middle of object*/,
> /*Comment in the middle of array*/,

This behavior is existing one, I did not change and just include them into test cases.
Both #75075 and #52408 should have be covered by test cases.

krwq commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:
```diff
@@ -7747,6 +7893,24 @@ private static string GetDatesExpectedString(bool prettyPrint, string keyString,
+            if (prettyPrint)
+            {
+                json.Flush();
+                streamWriter.Write(Enumerable.Repeat(' ', whitespaceCount).ToArray());
```
what about `new string(' ', whitespaceCount)`

lateapexearlyspeed replied:
Got it, thanks!

krwq commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:
```diff
@@ -7747,6 +7893,24 @@ private static string GetDatesExpectedString(bool prettyPrint, string keyString,
             return Encoding.UTF8.GetString(ms.ToArray());
         }
 
+        private static void CompensateWhitespaces(bool prettyPrint, JsonTextWriter json, TextWriter streamWriter, int whitespaceCount = 1)
```
honestly I'm not sure what's the intent of this method...

lateapexearlyspeed replied:
Previously we may need this helper method to instrument underlying stream to add additional newline and indented whitespaces during call newtonsoft.json methods because newtonsoft.json also did not do what proposal describes.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs:
```diff
@@ -541,12 +543,9 @@ private void WriteStartIndented(byte token)
             }
 
-            if (_tokenType != JsonTokenType.PropertyName)
+            if ((_tokenType != JsonTokenType.PropertyName && _tokenType != JsonTokenType.None) || _commentAfterNoneOrPropertyName)
```
Forget the exact syntax but could use pattern matching instead:
```suggestion
            if ((_tokenType is not JsonTokenType.PropertyName or is not JsonTokenType.None) || _commentAfterNoneOrPropertyName)
```

lateapexearlyspeed replied:
Fixed.

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:
```diff
@@ -4208,6 +4208,112 @@ private static string GetCommentExpectedString(bool prettyPrint)
+        public void WriteCommentsInObjectAsIndentedJson()
+        {
+            const string expectedIndentedJson =
+@"/*Comment at start of doc*/
```
You can use the new raw string literals feature for more readable JSON:

https://devblogs.microsoft.com/dotnet/csharp-11-preview-updates/#raw-string-literals

lateapexearlyspeed replied:
Changed.

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:
```diff
@@ -4208,6 +4208,112 @@ private static string GetCommentExpectedString(bool prettyPrint)
+]
+/*Comment at end of doc*/";
+
+            var options = new JsonWriterOptions { Indented = true, SkipValidation = false };
```
Why do we need to skip validation here? IIRC comments are still allowed with validation enabled.

lateapexearlyspeed replied:
It was enabling validation: `SkipValidation = false`.

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:
```diff
@@ -4208,6 +4208,112 @@ private static string GetCommentExpectedString(bool prettyPrint)
+        [Fact]
+        public void WriteCommentsInArrayAsIndentedJson()
+        {
+            const string expectedIndentedJson =
```
Thanks for adding string literal tests. For completeness, would it be possible to add an equivalent one with indentation disabled?

lateapexearlyspeed replied:
Added.

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:
```diff
@@ -4102,74 +4100,7 @@ public void WriteCommentsInArray(bool formatted, bool skipValidation, string com
-        }
-
-        [Theory]
-        [InlineData(true, true, "comment")]
```
It would be nice to keep these tests as well, complementing the string literal tests.

lateapexearlyspeed replied:
Added back.

eiriktsarpalis commented:
Could you try rebasing your branch? There appear to be merge conflicts

lateapexearlyspeed commented:
> Could you try rebasing your branch? There appear to be merge conflicts

Resolved, thanks !

krwq commented:
/azp run runtime

lateapexearlyspeed commented:
Hi @eiriktsarpalis @krwq, just kindly remind we had fixed or explained for all comments. Please check if anything missed to do :)

krwq commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:
```diff
@@ -23,6 +23,42 @@ public partial class Utf8JsonWriterTests
         private const int MaxUnescapedTokenSize = MaxEscapedTokenSize / MaxExpansionFactorWhileEscaping;  // 166_666_666 bytes
 
+        private const string ExpectedIndentedCommentJsonOfArray = """
+/*Comment at start of doc*/
```
Is there a test for writing comment twice in a row?

lateapexearlyspeed replied:
Added.
There was already test case '[WriteCommentsInvalidTextAllowed](https://github.com/dotnet/runtime/blob/e1ddbca5a008b4b15e8e6b309cbab12ba8b9b842/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs#L4154)' which wrote multiple comments in same line (so make sure behavior is not broken), but still added here for visible version. @krwq

krwq commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:
```diff
@@ -23,6 +23,42 @@ public partial class Utf8JsonWriterTests
         private const int MaxUnescapedTokenSize = MaxEscapedTokenSize / MaxExpansionFactorWhileEscaping;  // 166_666_666 bytes
 
+        private const string ExpectedIndentedCommentJsonOfArray = """
+/*Comment at start of doc*//*Comment in same line*/
+[
+  /*Comment as first array item*/
```
Why not like this:

```suggestion
/*Comment at start of doc*/
/*Comment in same line*/
[
  /*Comment as first array item*/
```

lateapexearlyspeed replied:
Changed.

krwq reviewed (approved):
LGTM, I have slight preference on putting extra newline between two comments but it's not blocking to me since it likely is not very common. Let me know if you're planning to fix this or not

krwq commented:
Any chance you could run perf tests on this change? There are changes made to the hot path so it would be useful to know if we regressed something or not and if yes how can we mitigate this

lateapexearlyspeed commented:
> Let me know if you're planning to fix this or not

Changed, please check.

> Any chance you could run perf tests

I would try to run perf test after deciding current json format is ok, please help confirm, thanks @krwq !

lateapexearlyspeed commented:
@krwq I run perf tests under [Utf8JsonWriter](https://github.com/dotnet/performance/tree/main/src/benchmarks/micro/libraries/System.Text.Json/Utf8JsonWriter), slower/faster result of perf test is not very stable, following is one sample:
dotnet run --base "C:\results\before" --diff "C:\results\after" --threshold 2%
summary:
better: 6, geomean: 1.046
worse: 4, geomean: 1.057
total diff: 10

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Tests.Writer.Perf_Strings.WriteStringsUtf8(Formatted: False, Sk |      1.08 |       5282410.64 |       5689104.17 |         |
| System.Text.Json.Tests.Writer.Perf_Guids.WriteGuids(Formatted: True, SkipValidat |      1.06 |       1934436.72 |       2054671.88 |         |
| System.Text.Json.Tests.Writer.Perf_Basic.WriteBasicUtf16(Formatted: True, SkipVa |      1.05 |       1285166.32 |       1352369.29 |         |
| System.Text.Json.Tests.Writer.Perf_Strings.WriteStringsUtf16(Formatted: False, S |      1.04 |      28113075.00 |      29187925.00 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Tests.Writer.Perf_Strings.WriteStringsUtf16(Formatted: True, Sk |      1.07 |       7557617.65 |       7081003.13 |         |
| System.Text.Json.Tests.Writer.Perf_Basic.WriteBasicUtf8(Formatted: True, SkipVal |      1.05 |           381.68 |           362.26 |         |
| System.Text.Json.Tests.Writer.Perf_Basic.WriteBasicUtf16(Formatted: False, SkipV |      1.05 |       1134185.27 |       1076607.92 |         |
| System.Text.Json.Tests.Writer.Perf_Basic.WriteBasicUtf16(Formatted: False, SkipV |      1.04 |       1028596.29 |        990623.24 |         |
| System.Text.Json.Tests.Writer.Perf_Base64.WriteByteArrayAsBase64_NoEscaping(Numb |      1.03 |           189.53 |           183.39 |         |
| System.Text.Json.Tests.Writer.Perf_Base64.WriteByteArrayAsBase64_HeavyEscaping(N |      1.03 |           183.19 |           178.23 |         |

And run perf test for "worst" case `System.Text.Json.Tests.Writer.Perf_Strings.WriteStringsUtf8` as above sample (note it is not always bad one), and found it has no obvious regression:
BenchmarkDotNet=v0.13.1.1603-nightly, OS=Windows 10.0.22621
12th Gen Intel Core i7-12800H, 1 CPU, 20 logical and 14 physical cores
.NET SDK=6.0.402
  [Host]     : .NET 6.0.10 (6.0.1022.47605), X64 RyuJIT
  Job-FHXNQJ : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT
  Job-HJTGLI : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:DebugType=portable,-bl:benchmarkdotnet.binlog  IterationTime=250.0000 ms
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1

|           Method |        Job |                 Toolchain | Formatted | SkipValidation |     Escaped |      Mean |     Error |    StdDev |    Median |       Min |       Max | Ratio | MannWhitney(2%) | RatioSD | Allocated |
|----------------- |----------- |-------------------------- |---------- |--------------- |------------ |----------:|----------:|----------:|----------:|----------:|----------:|------:|---------------- |--------:|----------:|
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |     False |          False |  AllEscaped | 24.862 ms | 0.4697 ms | 0.5025 ms | 24.809 ms | 24.186 ms | 25.897 ms |  0.99 |            Same |    0.03 |     288 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |     False |          False |  AllEscaped | 25.110 ms | 0.5516 ms | 0.6353 ms | 24.949 ms | 24.178 ms | 26.502 ms |  1.00 |            Base |    0.00 |     288 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |     False |          False |  OneEscaped |  5.549 ms | 0.1038 ms | 0.1019 ms |  5.560 ms |  5.339 ms |  5.746 ms |  1.00 |            Same |    0.03 |     135 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |     False |          False |  OneEscaped |  5.572 ms | 0.1085 ms | 0.1250 ms |  5.592 ms |  5.211 ms |  5.750 ms |  1.00 |            Base |    0.00 |     176 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |     False |          False | NoneEscaped |  3.123 ms | 0.0517 ms | 0.0483 ms |  3.141 ms |  3.040 ms |  3.180 ms |  1.00 |            Same |    0.03 |     124 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |     False |          False | NoneEscaped |  3.120 ms | 0.0617 ms | 0.0606 ms |  3.095 ms |  3.028 ms |  3.229 ms |  1.00 |            Base |    0.00 |     124 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |     False |           True |  AllEscaped | 24.777 ms | 0.5306 ms | 0.6110 ms | 24.507 ms | 23.961 ms | 25.910 ms |  0.99 |            Same |    0.04 |     288 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |     False |           True |  AllEscaped | 25.188 ms | 0.4922 ms | 0.5055 ms | 25.261 ms | 24.368 ms | 26.062 ms |  1.00 |            Base |    0.00 |     288 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |     False |           True |  OneEscaped |  5.390 ms | 0.1068 ms | 0.0999 ms |  5.390 ms |  5.237 ms |  5.533 ms |  1.02 |            Same |    0.03 |     140 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |     False |           True |  OneEscaped |  5.304 ms | 0.1031 ms | 0.1146 ms |  5.336 ms |  5.035 ms |  5.477 ms |  1.00 |            Base |    0.00 |     135 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |     False |           True | NoneEscaped |  3.068 ms | 0.0584 ms | 0.0573 ms |  3.089 ms |  2.961 ms |  3.132 ms |  0.99 |            Same |    0.02 |     124 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |     False |           True | NoneEscaped |  3.081 ms | 0.0349 ms | 0.0309 ms |  3.085 ms |  3.000 ms |  3.124 ms |  1.00 |            Base |    0.00 |     124 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |      True |          False |  AllEscaped | 24.923 ms | 0.3265 ms | 0.3054 ms | 24.892 ms | 24.620 ms | 25.546 ms |  0.99 |            Same |    0.03 |     288 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |      True |          False |  AllEscaped | 25.069 ms | 0.4837 ms | 0.5176 ms | 25.012 ms | 24.214 ms | 26.047 ms |  1.00 |            Base |    0.00 |     288 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |      True |          False |  OneEscaped |  5.655 ms | 0.0670 ms | 0.0627 ms |  5.671 ms |  5.481 ms |  5.742 ms |  1.00 |            Same |    0.02 |     135 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |      True |          False |  OneEscaped |  5.658 ms | 0.1129 ms | 0.1255 ms |  5.630 ms |  5.438 ms |  5.824 ms |  1.00 |            Base |    0.00 |     136 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |      True |          False | NoneEscaped |  3.325 ms | 0.0441 ms | 0.0391 ms |  3.322 ms |  3.244 ms |  3.392 ms |  0.97 |            Same |    0.02 |     125 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |      True |          False | NoneEscaped |  3.418 ms | 0.0603 ms | 0.0592 ms |  3.410 ms |  3.341 ms |  3.562 ms |  1.00 |            Base |    0.00 |     124 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |      True |           True |  AllEscaped | 24.961 ms | 0.3017 ms | 0.2356 ms | 24.953 ms | 24.644 ms | 25.529 ms |  0.98 |            Same |    0.02 |     288 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |      True |           True |  AllEscaped | 25.318 ms | 0.4664 ms | 0.4134 ms | 25.259 ms | 24.831 ms | 26.301 ms |  1.00 |            Base |    0.00 |     288 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |      True |           True |  OneEscaped |  5.746 ms | 0.0776 ms | 0.0726 ms |  5.731 ms |  5.581 ms |  5.857 ms |  1.00 |            Same |    0.02 |     136 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |      True |           True |  OneEscaped |  5.722 ms | 0.0883 ms | 0.0826 ms |  5.729 ms |  5.586 ms |  5.820 ms |  1.00 |            Base |    0.00 |     135 B |
|                  |            |                           |           |                |             |           |           |           |           |           |           |       |                 |         |           |
| WriteStringsUtf8 | Job-FHXNQJ |  \8.0.0-after\CoreRun.exe |      True |           True | NoneEscaped |  3.266 ms | 0.0370 ms | 0.0346 ms |  3.265 ms |  3.193 ms |  3.323 ms |  1.01 |            Same |    0.02 |     124 B |
| WriteStringsUtf8 | Job-HJTGLI | \8.0.0-before\CoreRun.exe |      True |           True | NoneEscaped |  3.229 ms | 0.0383 ms | 0.0359 ms |  3.242 ms |  3.139 ms |  3.273 ms |  1.00 |            Base |    0.00 |     128 B |

krwq commented on src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Comment.cs:
```diff
@@ -113,14 +117,13 @@ private void WriteCommentIndented(ReadOnlySpan<char> value)
             Span<byte> output = _memory.Span;
 
-            if (_tokenType != JsonTokenType.None)
+            if (_tokenType != JsonTokenType.None || _commentAfterNoneOrPropertyName)
```
optional suggestion: rename `_commentAfterNoneOrPropertyName` to `_forceNewLineAfterComment` - it took me a while to figure out what this code does - I think with the suggested name it makes it more obvious.

krwq commented:
/azp run runtime

lateapexearlyspeed committed:
Make WriteCommentValue overloads have same behavior
Squashed commit of the following:

commit e89bcfe2d8c3f45232f3d0b861be34ccede948f1
Author: lateapexearlyspeed <lateapexearlyspeed@163.com>
Date:   Wed Nov 9 11:56:41 2022 +0800

    Change comment behavior: add newline between comments.

commit 60d730ae151c9c2bffdaf7b6ca3cfa68180ba8b8
Author: lateapexearlyspeed <lateapexearlyspeed@163.com>
Date:   Thu Oct 27 15:36:43 2022 +0800

    Fix comment: test multiple comments in same line.

commit 977e76a298e52ad184adf78ab7172be99b97b9e2
Merge: 64b6be97fd4 9b9b0c95c67
Author: LateApexEarlySpeed <72254037+lateapexearlyspeed@users.noreply.github.com>
Date:   Fri Sep 23 12:56:01 2022 +0800

    Merge branch 'main' into lateapexearlyspeed-MakeWriteCommentValueOverloadSameBehavior

commit 64b6be97fd4c9b59fc601db8a0e6ddd8a8b97a7f
Author: lateapexearlyspeed <lateapexearlyspeed@163.com>
Date:   Thu Sep 22 10:39:32 2022 +0800

    Fix comment: keep newtonsoft.json asserting tests; add non-indented tests; use pattern match.

commit e3a772ac985801eb89abc028f5a71f2ad502cd55
Author: lateapexearlyspeed <lateapexearlyspeed@163.com>
Date:   Tue Sep 20 16:20:59 2022 +0800

    Inline expectedIndentedJson to local var.

commit 9e16463cc8826890e6fbeb115ba9a05867ee5017
Author: lateapexearlyspeed <lateapexearlyspeed@163.com>
Date:   Tue Sep 20 11:46:44 2022 +0800

    Fix comment: update test cases to use string literal to assert.

commit bebb09c0c9bc7b7b1c5076ab6dbcf1f5ec1d884a
Author: lateapexearlyspeed <lateapexearlyspeed@163.com>
Date:   Thu Sep 15 17:56:44 2022 +0800

    Add tests.

commit 2f9dbff3629f0b9b8b24f2bb2c32ec9c620a6fcd
Author: lateapexearlyspeed <lateapexearlyspeed@163.com>
Date:   Wed Sep 7 11:01:11 2022 +0800

    Add NewLine between comment and start of (object and array).

commit 9c1bb280f59e3755227747251a1a51fdca9eeab1
Author: lateapexearlyspeed <lateapexearlyspeed@163.com>
Date:   Mon Sep 5 15:11:57 2022 +0800

    Make WriteCommentValue() overloads have same behavior.

krwq commented:
(I force pushed again because git changed author to me so I changed you back as the author by hand)

krwq commented:
Thanks @lateapexearlyspeed!
